oda_api/api.py

```diff
@@ -1339,3 +1340,48 @@ def from_response_json(cls, res_json, instrument, product):
                 p.meta_data = p.meta
 
         return d
+
+class ProgressReporter(object):
+    """
+    The class allows to report task progress to end user
+    """
+    def __init__(self):
+        self._callback = None
+        callback_file = ".oda_api_callback"  # perhaps it would be better to define this constant in a common lib
+        if not os.path.isfile(callback_file):
+            return
+        with open(callback_file, 'r') as file:
+            self._callback = file.read().strip()
+
+    @property
+    def enabled(self):
+        return self._callback is not None
+
+    def report_progress(self, stage: str=None, progress: int=50, substage: str=None, subprogress: int=None, message:str=None):
+        """
+        Report progress via callback URL
+        :param stage: current stage description string
+        :param progress: current stage progress in %
+        :param substage: current substage description string
+        :param subprogress: current substage progress in %
+        :param message: message to pass
+        """
+        callback_payload = {k: str(v) for k, v in locals().items() if v is not None and k != 'self'}
```

Comment:
@burnout87 @volodymyrss I'm not sure, is it needed for dispatcher that `"action": "progress"` is in the callback payload?

Reply:
Do you want to refer to the code where this is parsed?

Reply:
dispatcher uses it here: https://github.com/oda-hub/dispatcher-app/blob/5b8c558a8d107d94774813f00402ca92ebeca2d8/cdci_data_analysis/flask_app/dispatcher_query.py#L1069-L1074

dispatcher needs to know that, and will it for email/matrix messages

Reply:
"progress" will be shown in frontend and API response. People do distinguish "progress" from "unknown", visually :) This is for people.

Reply:
it does not, but it will appear within the email/matrix message, but of course it will not disrupt the overall execution of the job

Reply:
This whole PR is to help user understand what's happening, we do not want to have user see numerous "unknown" word as sign of progress :)